rpc/wrtc_rate_limiter.go

```diff
@@ -0,0 +1,180 @@
+package rpc
+
+import (
+	"context"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"go.viam.com/utils"
+	mongoutils "go.viam.com/utils/mongo"
+	"go.viam.com/utils/perf/statz"
+	"go.viam.com/utils/perf/statz/units"
+)
+
+func init() {
+	mongoutils.MustRegisterNamespace(&mongodbWebRTCCallQueueDBName, &MongoDBRateLimiterCollName)
+}
+
+var rateLimitDenials = statz.NewCounter1[string]("signaling/rate_limits_denials", statz.MetricConfig{
+	Description: "Total number of requests rate limited.",
+	Unit:        units.Dimensionless,
+	Labels: []statz.Label{
+		{Name: "key", Description: "Method and auth ID of the client being rate limited."},
+	},
+})
+
+// Database configuration and collection names for MongoDB rate limiter.
+var (
+	MongoDBRateLimiterCollName     = "rate_limiter"
+	mongodbWebRCRateLimiterTTLName = "rate_limit_expire"
+)
+
+// RateLimitConfig specifies the configuration for rate limiting in terms of maximum requests allowed in a given time window.
+type RateLimitConfig struct {
+	MaxRequests int
+	Window      time.Duration
+}
+
+// A MongoDBRateLimiter is a MongoDB implementation of a continuous sliding rate limiter designed to be used for
+// multi-node, distributed deployments.
+type MongoDBRateLimiter struct {
+	rateLimitColl *mongo.Collection
+	config        RateLimitConfig
+	logger        utils.ZapCompatibleLogger
+}
+
+// NewMongoDBRateLimiter returns a new MongoDB based rate limiter where requests are allowed or denied based on how many
+// requests have been made by a specific key (e.g., method + auth ID) within a certain time window specified by the limit
+// provided by the config.
+func NewMongoDBRateLimiter(
+	ctx context.Context,
+	client *mongo.Client,
+	config RateLimitConfig,
+	logger utils.ZapCompatibleLogger,
+) (*MongoDBRateLimiter, error) {
+	rateLimitColl := client.Database(mongodbWebRTCCallQueueDBName).Collection(MongoDBRateLimiterCollName)
+
+	ttlSeconds := int32((2 * config.Window).Seconds())
```

Comment:
I'm guessing this is from my comment yesterday about how we could use ttl seconds rather than continually updating a field for the expiration. Sorry for flip-flopping on this but I don't think this is actually the way to go, simply because 1) this will involve constantly updating a value anyway and 2) changing the window will mean rebuilding this index if we do it this way. Also since we're constantly updating this field `created_at` is a bit untrue as a name. I would suggest

1) change this (back to?) `expires_at'
2) on each update, set it to '$NOW' + the window size
3) set the ttl index to delete it w/o any `expireAfterSeconds` option so it is deleted as soon as the expiry time hits (and the background thread gets around to it)